README.md
# basex

Convert data to any base (base32, base64, base58, etc) using an alphabet set. Supports disabling padding.
The advantage of this package is that it allows converting data bytes to base encoded bytes. 

# To base

```dart
import 'package:ninja_basex/ninja_basex.dart';
import 'package:ninja/ninja.dart';

void main() {
  {
    final out = List.filled(4, 0);
    pack('Man'.codeUnits, 64, 63, out, 0);
    print(out);
    // => [46, 0, 0, 0]
  }

  {
    final out = toBaseBytes('Man'.codeUnits, 64);
    print(out);
    // => [19, 22, 5, 46]
  }

  {
    final out = toBaseBytes('Ma'.codeUnits, 64);
    print(out);
    // => [19, 22, 4]
  }

  {
    final out = toBaseBytes('Ma'.codeUnits, 64, padding: 61);
    print(out);
    // => [19, 22, 4, 61]
  }

  {
    final out = toBase('Man'.codeUnits, base64Alphabet);
    print(out);
    // => TWFu
  }

  {
    final out = toBase('Ma'.codeUnits, base64Alphabet);
    print(out);
    // => TWE=
  }
  
  {
    final bint = BigInt.parse('751e76e8199196d454941c45d1b3a323f1433bd6', radix: 16);
    final input = bigIntToBytes(bint);
    final out = toBaseBytes(input, 32);
    print(out);
    // => [14, 20, 15, 7, 13, 26, 0, 25, 18, 6, 11, 13, 8, 21, 4, 20, 3, 17, 2, 29, 3, 12, 29, 3, 4, 15, 24, 20, 6, 14, 30, 22]
  }
}
```

# From bytes

```dart
import 'package:ninja_basex/ninja_basex.dart';

void main() {
  final bytes = fromBase(
      'RVIHETOST35MQXOPH66OL2WX6IJQQM54FIQMNKTKZOLCFS72OMBK6762BW6U74VB3RCZTE5J6533BE2FZRLJ3M3ZK37ZFW7VVDDBEDPVQICZ4OS3CAWACOTFX6GPTCCR2DOCQ6PET2KQNFNEZGMTQPJVNA',
      base32Alphabet);
  print(bytes);
  print(bytes.map((e) => e.toRadixString(16)).toList());
}
```

CHANGELOG.md
## 1.0.3

+ Added `fromBase` and `fromBaseBytes` to convert from base encoded string and bytes

## 1.0.2

+ README

## 1.0.1

+ Returned list is now growable

## 1.0.0

+ toBaseBytes
+ toBase

test/to_test.dart
import 'package:ninja_basex/ninja_basex.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {});
  });
}


test/basex_test.dart
import 'package:ninja_basex/ninja_basex.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {

    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {
    });
  });
}


lib/src/to.dart
import 'dart:math';

/// Converts given data bytes into base encoded string
String toBase(Iterable<int> input, String alphabet, {int? padding = 61}) {
  final bytes = toBaseBytes(input, alphabet.codeUnits.length, padding: padding);
  return baseBytesToString(bytes, alphabet, padding: padding);
}

/// Converts base encoded bytes into base encoded string
String baseBytesToString(List<int> bytes, String alphabet,
    {int? padding = 61}) {
  final sb = StringBuffer();
  for (int i = 0; i < bytes.length; i++) {
    final e = bytes[i];
    sb.write(String.fromCharCode(
        padding == null || e != padding ? alphabet.codeUnitAt(e) : padding));
  }
  return sb.toString();
}

/// Converts given data bytes into base bytes
List<int> toBaseBytes(Iterable<int> input, int base, {int? padding}) {
  int baseBit = (base - 1).bitLength;
  int mask = (pow(2, baseBit) - 1).toInt();
  int inpLength = input.length;
  int outputBlockLength = baseBit;
  while (outputBlockLength % 8 != 0) {
    outputBlockLength += baseBit;
  }
  int inputBlockLength = outputBlockLength ~/ 8;
  outputBlockLength ~/= baseBit;

  int count;
  if (padding == null) {
    int inputBitLength = inpLength * 8;
    count = (inputBitLength / baseBit).ceil();
  } else {
    count =
        (inpLength / inputBlockLength).ceil() * inputBlockLength * 8 ~/ baseBit;
  }

  int inputPointer = 0;
  int outputPointer = 0;

  final out = List<int>.filled(count, 0, growable: true);
  while (inpLength - inputPointer >= inputBlockLength) {
    pack(input.skip(inputPointer).take(inputBlockLength), baseBit, mask, out,
        outputPointer);
    inputPointer += inputBlockLength;
    outputPointer += outputBlockLength;
  }
  if (inpLength != inputPointer) {
    packPartial(input.skip(inputPointer), baseBit, mask, outputBlockLength, out,
        outputPointer,
        padding: padding);
  }
  return out;
}

void pack(
    Iterable<int> input, int base, int mask, List<int> out, int outStart) {
  int v = 0;
  for (int byte in input) {
    v <<= 8;
    v |= byte;
  }
  int count = ((input.length * 8) / base).ceil();
  for (int i = 0; i < count; i++) {
    int j = count - 1 - i;
    out[outStart++] = (v >> j * base) & mask;
  }
}

void packPartial(Iterable<int> input, int base, int mask, int outputBlockLength,
    List<int> out, int outStart,
    {int? padding}) {
  int v = 0;
  for (int byte in input) {
    v <<= 8;
    v |= byte;
  }

  int inputBitLength = input.length * 8;
  int dataCount = (inputBitLength / base).ceil();
  int outputBitCount = dataCount * base;
  int correction = outputBitCount - inputBitLength;
  v = v << correction;
  for (int i = 0; i < dataCount; i++) {
    int j = dataCount - 1 - i;
    out[outStart++] = (v >> j * base) & mask;
  }
  if (padding != null) {
    for (int i = dataCount; i < outputBlockLength; i++) {
      out[outStart++] = padding;
    }
  }
}


lib/src/from.dart
import 'dart:typed_data';

import 'package:ninja_basex/src/util.dart';

/// Converts base bytes to data bytes
List<int> fromBaseBytes(List<int> input, int base) {
  int baseBit = (base - 1).bitLength;
  final outLenInBits = input.length * baseBit;
  int shift = (8 - (outLenInBits % 8)) % 8;
  BigInt v = BigInt.zero;
  for (int i = 0; i < input.length; i++) {
    v <<= baseBit;
    v |= BigInt.from(input[i] & byteMask);
  }
  v <<= shift;
  final outLenInBytes = (outLenInBits / 8).ceil();
  final ret = Uint8List(outLenInBytes);
  for (int i = 0; i < outLenInBytes; i++) {
    final byte = (v & byteMaskBigInt).toInt();
    ret[ret.length - i - 1] = byte;
    v >>= 8;
  }
  return ret;
}

/// Converts given base encoded string to data bytes
List<int> fromBase(String input, String alphabet, {int padding = 61}) {
  final ret = input.codeUnits
      .map((e) => e == padding ? -2 : alphabet.indexOf(String.fromCharCode(e)))
      .where((e) => e != -2)
      .toList();
  if (ret.contains(-1)) {
    throw Exception('invalid input');
  }
  return fromBaseBytes(ret, alphabet.codeUnits.length);
}


example/basex_example.dart
import 'package:ninja_basex/ninja_basex.dart';
import 'package:ninja/ninja.dart';

void main() {
  {
    final out = List.filled(4, 0);
    pack('Man'.codeUnits, 64, 63, out, 0);
    print(out);
    // => [46, 0, 0, 0]
  }

  {
    final out = toBaseBytes('Man'.codeUnits, 64);
    print(out);
    // => [19, 22, 5, 46]
  }

  {
    final out = toBaseBytes('Ma'.codeUnits, 64);
    print(out);
    // => [19, 22, 4]
  }

  {
    final out = toBaseBytes('Ma'.codeUnits, 64, padding: 61);
    print(out);
    // => [19, 22, 4, 61]
  }

  {
    final out = toBase('Man'.codeUnits, base64Alphabet);
    print(out);
    // => TWFu
  }

  {
    final out = toBase('Ma'.codeUnits, base64Alphabet);
    print(out);
    // => TWE=
  }
  
  {
    final bint = BigInt.parse('751e76e8199196d454941c45d1b3a323f1433bd6', radix: 16);
    final input = bigIntToBytes(bint);
    final out = toBaseBytes(input, 32);
    print(out);
    // => [14, 20, 15, 7, 13, 26, 0, 25, 18, 6, 11, 13, 8, 21, 4, 20, 3, 17, 2, 29, 3, 12, 29, 3, 4, 15, 24, 20, 6, 14, 30, 22]
  }
}
